### PR TITLE
Integrate IM event with server

### PR DESCRIPTION
**Pull request by yunhanw-google**
#### Problem
No support for IM Event Server Integration

#### Change overview
-- Port change from #12059
-- Hook LogEvent with EventManagement, and make necessary change in EventManagement and EventLogger class.
-- Add TestCluster Event Integration.

#### Testing
Update python test to issue IM command which can instruct server to generate event, then issue read request to retrieve event from server. Then issue another command to generate another event, and check again via read interaction

### CODE REVIEW

**Review by yufengwangca — APPROVED** [review]

**Review by jmartinez-silabs — APPROVED** [review]

File: src/app/EventLogging.h [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    const T & mEventData;
+};
+
+template <typename T>
+CHIP_ERROR LogEvent(const T & aEventData, EndpointId aEndpoint, EventOptions aEventOptions, EventNumber & aEventNumber)
+{
+    EventLogger<T> eventData(aEventData);
+    ConcreteEventPath path(aEndpoint, aEventData.GetClusterId(), aEventData.GetEventId());
+    // log the actual event
+    aEventNumber              = 0;
```
Comment (yufengwangca) [inline src/app/EventLogging.h:50]: The eventNumber need to be an monotonically increasing unique number

Comment (yunhanw-google) [reply src/app/EventLogging.h:50]: Yes, it has been passed in logEvent api in event management,  and set it inside, so it is out Parma, before pass in, I initialize it as 0

File: src/app/clusters/test-cluster-server/test-cluster-server.cpp [hunk 2]
```diff
@@ -443,9 +444,28 @@ bool emberAfTestClusterClusterTestListStructArgumentRequestCallback(
...
+bool emberAfTestClusterClusterTestEmitTestEventRequestCallback(
+    CommandHandler * commandObj, const ConcreteCommandPath & commandPath,
+    const Commands::TestEmitTestEventRequest::DecodableType & commandData)
+{
+    Commands::TestEmitTestEventResponse::Type responseData;
+    Structs::SimpleStruct::Type arg4;
+    DataModel::List<const Structs::SimpleStruct::Type> arg5;
+    DataModel::List<const SimpleEnum> arg6;
+    EventOptions eventOptions;
+    Events::TestEvent::Type event{ commandData.arg1, commandData.arg2, commandData.arg3, arg4, arg5, arg6 };
```
Comment (robszewczyk) [inline src/app/clusters/test-cluster-server/test-cluster-server.cpp:456]:
```suggestion
    // TODO:  Add code to pull arg4, arg5 and arg6 from the arguments of the command
    Events::TestEvent::Type event{ commandData.arg1, commandData.arg2, commandData.arg3, arg4, arg5, arg6 };
```

**Review by robszewczyk — APPROVED** [review]

File: src/app/clusters/test-cluster-server/test-cluster-server.cpp (hunk 2 above)
Comment (yunhanw-google) [reply src/app/clusters/test-cluster-server/test-cluster-server.cpp:458]: done, thanks

File: src/app/EventLogging.h (hunk 1 above)
Comment (yunhanw-google) [reply src/app/EventLogging.h:70]: updated the doxygen on this part, thanks

Comment (yunhanw-google) [reply src/app/EventLogging.h]: removed this line, thanks

**Review by chrisdecenzo — APPROVED** [review]

**Review by jelderton — APPROVED** [review]

**Pull request merged**